File: napmo/grids/multi_grid.py
# file: multi_grid.py
# nAPMO package
# Copyright (c) 2020, Edwin Fernando Posada
# All rights reserved.
# Version: 1.0
# fernando.posada@temple.edu

import numpy as np

import napmo

class MultiGrid(object):
	"""
	This class handles multiple BeckeGrids. 

	Sometimes it is important to know the common points between molecular grids
	specifically in the case of multi-species calculations.

	Args:
		nspecies (int): Total number of unique species in the system

	"""
	def __init__(self, nspecies):
		super(MultiGrid, self).__init__()
		
		assert isinstance(nspecies, int)

		self._grids = {}
		self._this = napmo.cext.MultiGrid_new(nspecies)

	def add_grid(self, grid):
		"""
		Adds a grid to the class

		Args:
			grid (BeckeGrid): Grid to be added, must be of the type of ``BeckeGrid``.
		"""
		assert isinstance(grid, napmo.BeckeGrid)
		
		# append grid
		grid_id = napmo.cext.MultiGrid_add_grid(self._this, grid._this)
		self._grids[grid._symbol] = {"id": grid_id, "obj": grid}


	def get_grid(self, symbol="", gid=-1):
		"""
		Returns the grid with a specific symbol or ID

		Args:
			symbol (str): Symbol of the desired grid
			gid (int): ID of the grid

		Notes: 
			One of the two arguments have to be provided, ``symbol`` has priority over ``gid``
			if both are provided.
		"""
		assert isinstance(symbol, str)
		assert isinstance(gid, int)

		# Symbol case
		aux = self._grids.get(symbol, None)
		if aux:
			return aux.get('obj', None)

		# Id case
		for grid in self._grids.values():
			if grid.get('id', -2) == gid:
				return grid.get('obj', None)

		# Default case
		return aux

	def get_grid_id(self, symbol):
		"""
		Returns the ID for a given grid

		Args:
			symbol (str): Symbol of the desired grid
		"""
		assert isinstance(symbol, str)

		aux = self._grids.get(symbol, None)

		if aux:
			return aux.get('id', None)
		return aux		

	def get_common_points(self, symbol_a, symbol_b):
		"""
		Returns the points with same coordinates between two grids.

		Args:
			symbol_a (str): Symbol of the species ``a``.
			symbol_b (str): Symbol of the species ``b``.

		Returns:
			output (ndarray): Array with the index in which the cartesian points are the same.
			The size of the array is the number of common points ``n`` with the shape ``(n, 2)``
			in which ``[:,0]`` corresponds to de index for ``a`` while ``[:,1]`` are the index for ``b``.
		"""
		assert isinstance(symbol_a, str)
		assert isinstance(symbol_b, str)

		idx_a = self.get_grid_id(symbol_a)
		idx_b = self.get_grid_id(symbol_b)

		# Is the same grid
		if idx_a == idx_b:
			return np.arange(self.get_grid(symbol_a).size)
		
		# Uses C function
		if idx_a != None and idx_b != None:
			ptr = napmo.cext.MultiGrid_get_common_index(self._this, idx_a, idx_b)
			size = napmo.cext.MultiGrid_get_common_index_size(self._this, idx_a, idx_b)
			return np.ctypeslib.as_array(ptr, shape=(size,2))

		# No grids for such species (or one of them)
		else:
			return None

	def show(self):
		"""
		Shown information about the object
		"""
		print("\nNumber of Grids: ", self.ngrids)
		print("Object Information: ", self._grids)

	@property
	def ngrids(self):
		"""
		Returns the number of the grids in the object
		"""
		return napmo.cext.MultiGrid_get_ngrids(self._this)

	@property
	def nspecies(self):
		"""
		Returns the number of species used to create the object
		"""
		return napmo.cext.MultiGrid_get_nspecies(self._this)


	
	

	

File: src/grid/multi_grid.cpp
/*
file: multi_grid.cpp
nAPMO package
Copyright (c) 2021, Edwin Fernando Posada
All rights reserved.
Version: 2.0
fernando.posada@temple.edu
*/

#include "multi_grid.h"

#include <stdio.h>

MultiGrid::MultiGrid(const int ns) {
  nspecies = ns;
  ngrids = 0;
  molgrid.reserve(nspecies);

  int aux = nspecies * (nspecies - 1) / 2;
  if (aux > 0) {
    common_index.reserve(aux);
    for (int i = 0; i < aux; ++i) {
      index_calculated.push_back(false);
      index_size.push_back(-1);
    }
  }
}

int MultiGrid::add_grid(BeckeGrid *grid) {
  molgrid.push_back(grid);
  ngrids++;

  // return index
  return ngrids - 1;
}

int *MultiGrid::get_common_index(const int id_grid_a, const int id_grid_b) {
  int idx = get_index(id_grid_a, id_grid_b);
  int id_a = id_grid_a;
  int id_b = id_grid_b;

  if (!index_calculated[idx]) {
    unsigned int size_a = molgrid[id_a]->get_size();
    unsigned int size_b = molgrid[id_b]->get_size();

    // B has to be a subset of A (A is always >= B)
    if (size_b > size_a) {
      std::swap(size_a, size_b);
      std::swap(id_a, id_b);
    }

    double *points_a = molgrid[id_a]->get_points();
    double *points_b = molgrid[id_b]->get_points();

    unsigned int counter = 0;
    unsigned int idx_i = 0;
    unsigned int idx_j = 0;

    std::vector<unsigned int> aux;

    for (unsigned int i = 0; i < size_a; ++i) {
      idx_i = i * 3;
      for (unsigned int j = 0; j < size_b; ++j) {
        idx_j = j * 3;
        if (points_a[idx_i + 0] == points_b[idx_j + 0] &&
            points_a[idx_i + 1] == points_b[idx_j + 1] &&
            points_a[idx_i + 2] == points_b[idx_j + 2]) {
          aux.push_back(i);
          aux.push_back(j);
          counter++;
        }
      }
    }

    int *idx_all = new int[counter * 2];
    for (unsigned int i = 0; i < counter * 2; ++i) {
      idx_all[i] = aux[i];
    }

    common_index.push_back(idx_all);
    index_calculated[idx] = true;
    index_size[idx] = counter;
  }

  return common_index[idx];
}

int MultiGrid::get_common_index_size(const int id_grid_a, const int id_grid_b) {
  int idx = get_index(id_grid_a, id_grid_b);
  return index_size[idx];
}

int MultiGrid::get_index(const int i, const int j) {
  int idx = -1;

  if (i > j) {
    idx = (3 * j) - ((j - 1) * (j / 2)) + (i - j) - 1;
  } else if (j > i) {
    idx = (3 * i) - ((i - 1) * (i / 2)) + (j - i) - 1;
  }

  return idx;
}

/*
Python wrapper
*/

MultiGrid *MultiGrid_new(int ns) { return new MultiGrid(ns); }

void MultiGrid_del(MultiGrid *multi_grid) { multi_grid->~MultiGrid(); }

int *MultiGrid_get_common_index(MultiGrid *multi_grid, int id_grid_a,
                                int id_grid_b) {
  return multi_grid->get_common_index(id_grid_a, id_grid_b);
}

int MultiGrid_get_common_index_size(MultiGrid *multi_grid, int id_grid_a,
                                    int id_grid_b) {
  return multi_grid->get_common_index_size(id_grid_a, id_grid_b);
}

int MultiGrid_add_grid(MultiGrid *multi_grid, BeckeGrid *grid) {
  return multi_grid->add_grid(grid);
}

int MultiGrid_get_nspecies(MultiGrid *multi_grid) {
  return multi_grid->get_nspecies();
}

int MultiGrid_get_ngrids(MultiGrid *multi_grid) {
  return multi_grid->get_ngrids();
}